classes/SpecialMission.php

```diff
@@ -784,6 +802,64 @@ private function calcHealthLost(int|float $base_hp_lost_percent): int {
         return floor(($hp_lost_percent / 100) * $this->player->max_health);
     }
 
+    public function downgradeMissionRank(int $mission_rank): int {
+        return match ($mission_rank) {
+            Mission::RANK_S => Mission::RANK_A,
+            Mission::RANK_A => Mission::RANK_B,
+            Mission::RANK_B => Mission::RANK_C,
+            default => Mission::RANK_D,
+        };
+    }
+
+    public function upgradeMissionRank(int $mission_rank): int {
+        return match ($mission_rank) {
+            Mission::RANK_A => Mission::RANK_S,
+            Mission::RANK_B => Mission::RANK_A,
+            Mission::RANK_C => Mission::RANK_B,
+            default => Mission::RANK_C,
+        };
+    }
+    public function getMissionRank(): int {
+        $mission_rank = match ($this->player->rank_num) {
+            5 => Mission::RANK_A,
+            4 => Mission::RANK_B,
+            3 => Mission::RANK_C,
+            default => Mission::RANK_D,
+        };
+
+        // Difficulty downgrade
+        if($this->difficulty == self::DIFFICULTY_EASY) {
+            $mission_rank = $this->downgradeMissionRank($mission_rank);
+        }
+        // Difficulty upgrade
+        if(in_array($this->difficulty, self::$UPGRADE_DIFFICULTIES)) {
+            $mission_rank = $this->upgradeMissionRank($mission_rank);
+        }
+
+        // Upgrade level for capped Chuunin
+        if($this->player->rank_num == 3 && $this->player->level >= 50) {
+            // This is intended to be INCLUSIVE w/difficulty below
+            $mission_rank = $this->upgradeMissionRank($mission_rank);
+            // Downgrade based on easy & normal difficulties (intended to STACK with easy downgrade above)
+            if(in_array($this->difficulty, [self::DIFFICULTY_NORMAL, self::DIFFICULTY_EASY])) {
+                $mission_rank = $this->downgradeMissionRank($mission_rank);
+            }
+        }
+        // Jonin level upgrade, begin at level 90
+        if($this->player->rank_num == 4 && $this->player->level >= 90) {
+            $mission_rank = $this->upgradeMissionRank($mission_rank);
+        }
+        // Jonin level downgrade based on NORMAL difficulty (intended to STACK with easy downgrade above)
+        if($this->player->rank_num == 4 && $this->player->level >= 75 && in_array($this->difficulty, [self::DIFFICULTY_EASY, self::DIFFICULTY_NORMAL])) {
+            $mission_rank = $this->downgradeMissionRank($mission_rank);
+        }
```

Comment:
This logic gets a little weird and pushes me towards - should we do the level-based system where we assign a level to each SM and then a level range for each mission rank? I think that's easier to understand

Config array example

```php
const SPECIAL_MISSION_LEVELS = [
   // AS
   1 => [
       self::DIFFICULTY_EASY = 1,
       self::DIFFICULTY_MEDIUM => 3,
       self::DIFFICULTY_HARD => 6,
       self::DIFFICULTY_NIGHTMARE = 9
   ] 
];
```

this also allows us to dynamically tune difficulty to say "this level is when you should be able to clear this mission"

Reply:
Love how clean this is... my brain over-complicates things sometimes ~.~

Reply:
Don't look at early WoB code, it was pretty gnarly 😅